includes/class-events-store.php

```diff
@@ -0,0 +1,592 @@
+				break;
+			}
+
+			$page++;
+
+			// Something's probably wrong if a site has more than 1,500 pending cron actions
+			if ( $page > 15 ) {
+				do_action( 'a8c_cron_control_stopped_runaway_cron_option_rebuild' );
+				break;
+			}
+
+			// Loop through results and built output Core expects
+			if ( ! empty( $jobs_posts ) ) {
+				foreach ( $jobs_posts as $jobs_post ) {
+					// Alias event timestamp
+					$timestamp = $jobs_post->timestamp;
+
+					// If timestamp is invalid, event is removed to let its source fix it
+					if ( $timestamp <= 0 ) {
+						$this->mark_job_record_completed( $jobs_post->ID );
+						continue;
+					}
+
+					// Basic arguments to add a job to the array format Core expects
+					$action   = $jobs_post->action;
+					$instance = $jobs_post->instance;
+
+					// Populate remaining job data
+					$cron_array[ $timestamp ][ $action ][ $instance ] = array(
+						'schedule' => $jobs_post->schedule,
+						'args'     => $jobs_post->args,
+						'interval' => 0,
+					);
+
+					if ( isset( $jobs_post->interval ) ) {
+						$cron_array[ $timestamp ][ $action ][ $instance ]['interval'] = $jobs_post->interval;
+					}
+				}
+
+				// No need to keep looping if there were fewer events than we asked for
+				if ( count( $jobs_posts ) < $quantity ) {
+					break;
+				}
+			}
+		} while( true );
+
+		// Re-sort the array just as Core does when events are scheduled
+		// Ensures events are sorted chronologically
+		uksort( $cron_array, 'strnatcasecmp' );
+
+		// Cache the results, bearing in mind that they won't be used during unscheduling events
+		wp_cache_set( self::CACHE_KEY, $cron_array, null, 1 * \HOUR_IN_SECONDS );
+
+		return $cron_array;
+	}
+
+	/**
+	 * Handle requests to update the cron option
+	 *
+	 * By returning $old_value, `cron` option won't be updated
+	 */
+	public function update_option( $new_value, $old_value ) {
+		// Find changes to record
+		$new_events     = $this->find_cron_array_differences( $new_value, $old_value );
+		$deleted_events = $this->find_cron_array_differences( $old_value, $new_value );
+
+		// Add/update new events
+		foreach ( $new_events as $new_event ) {
+			$job_id = $this->job_exists( $new_event['timestamp'], $new_event['action'], $new_event['instance'], true );
+
+			if ( 0 === $job_id ) {
+				$job_id = null;
+			}
+
+			$this->create_or_update_job( $new_event['timestamp'], $new_event['action'], $new_event['args'], $job_id, false );
+		}
+
+		// Mark deleted entries for removal
+		foreach ( $deleted_events as $deleted_event ) {
+			$this->mark_job_completed( $deleted_event['timestamp'], $deleted_event['action'], $deleted_event['instance'], false );
+		}
+
+		$this->flush_internal_caches();
+
+		return $old_value;
+	}
+
+	/**
+	 * When an entry exists, don't try to create it again
+	 */
+	public function block_creation_if_job_exists( $job ) {
+		$instance = md5( maybe_serialize( $job->args ) );
+		if ( $this->job_exists( $job->timestamp, $job->hook, $instance ) ) {
```

Comment:
Alternatively, following Josh's suggestion, if the method returns a `bool`, we could just do `return $this->job_exists( ... )`